File: server/service.ts
import type { Agent, Project, Run, Task } from '../shared/contracts.ts'
import type { Config } from './config.ts'
import type { Store } from './store.ts'
import { execFile } from 'node:child_process'
import { randomUUID } from 'node:crypto'
import { promisify } from 'node:util'
import { CronExpressionParser } from 'cron-parser'
import { agentInput, MAIN_AGENT_ID, projectInput, taskInput } from '../shared/contracts.ts'
import { Chats } from './chats.ts'
import { CodexAccounts } from './codex-accounts.ts'
import { AppError, requireValue } from './errors.ts'
import { McpConnections } from './mcp-connections.ts'
import { workspaceDirectory } from './paths.ts'
import { allowedProjects, policy, runProjects, taskProjects, validateAccess } from './policy.ts'
import { Skills } from './skills.ts'

const exec = promisify(execFile)
async function projectOrigin(directory: string) {
  try {
    const { stdout } = await exec('git', ['-C', directory, 'config', '--get', 'remote.origin.url'], { timeout: 3000, maxBuffer: 8000 })
    const remote = stdout.trim()
    const scp = remote.match(/^(?:[\w.-]+@)?([a-z0-9.-]+):([\w./-]+)$/i)
    if (scp)
      return `${scp[1]}:${scp[2]}`
    const url = new URL(remote)
    if (!['https:', 'http:', 'ssh:', 'git:'].includes(url.protocol))
      return ''
    return `${url.protocol}//${url.host}${url.pathname}`
  }
  catch {
    return ''
  }
}

export function nextOccurrences(
  cron: string,
  timezone: string,
  now = Date.now(),
  count = 3,
) {
  try {
    if (cron.trim().split(/\s+/).length !== 5)
      throw new Error('Cron requires five fields')
    new Intl.DateTimeFormat('en', { timeZone: timezone }).resolvedOptions()
    const expression = CronExpressionParser.parse(cron, {
      tz: timezone,
      currentDate: new Date(now),
    })
    return Array.from({ length: count }, () => expression.next().getTime())
  }
  catch {
    throw new AppError(
      400,
      'Enter a valid five-field cron expression and IANA timezone.',
    )
  }
}
export class Service {
  readonly chats = new Chats(this)
  skills: Skills
  mcps: McpConnections
  accounts: CodexAccounts
  constructor(
    public store: Store,
    public config: Config,
  ) {
    this.mcps = new McpConnections(store, config)
    this.accounts = new CodexAccounts(store, config)
    this.skills = new Skills(config.home, config.workspaceRoots)
    this.store.transaction(() => {
      for (const agent of store.list('agents')) {
        if (!agent.access || !Object.hasOwn(agent.access, 'mcps'))
          store.put('agents', { ...agent, access: policy(agent) })
      }
      if (!store.get('agents', MAIN_AGENT_ID))
        store.put('agents', { ...agentInput.parse({ name: 'Main agent', description: 'Your default agent, with access to every registered project, skill, and shared connection.' }), id: MAIN_AGENT_ID, createdAt: Date.now() })
    })
  }

  agent(value: unknown, existingId?: string) {
    const input = agentInput.parse(value)
    const existing = existingId ? requireValue(this.store.get('agents', existingId)) : undefined
    if (existing && value && typeof value === 'object' && !Object.hasOwn(value, 'access'))
      input.access = policy(existing)
    const item: Agent = {
      ...input,
      id: existingId ?? randomUUID(),
      createdAt:
        this.store.get('agents', existingId ?? '')?.createdAt ?? Date.now(),
    }
    validateAccess(item)
    for (const projectId of item.access.projects ?? [])
      requireValue(this.store.get('projects', projectId), 'Allowed project not found')
    if (existingId === MAIN_AGENT_ID && (item.access.projects !== null || item.access.skills !== null || !item.access.github || item.access.mcps !== null || Object.keys(item.access.mcpTools).length > 0))
      throw new AppError(400, 'The main agent always has access to all resources. Create another agent for restricted access.')
    for (const id of item.access.mcps ?? []) this.mcps.get(id)
    for (const id of Object.keys(item.access.mcpTools)) {
      this.mcps.get(id)
      if (item.access.mcps !== null && !item.access.mcps.includes(id))
        throw new AppError(400, 'Tool permissions require access to the MCP connection.')
    }
    this.store.put('agents', item)
    this.store.audit('agent.saved', { id: item.id })
    return item
  }

  async project(value: unknown, existingId?: string) {
    const input = projectInput.parse(value)
    const actual = await workspaceDirectory(
      input.path,
      this.config.workspaceRoots,
    )
    if (existingId)
      requireValue(this.store.get('projects', existingId))
    const item: Project = {
      ...input,
      path: actual,
      origin: await projectOrigin(actual),
      id: existingId ?? randomUUID(),
      createdAt:
        this.store.get('projects', existingId ?? '')?.createdAt ?? Date.now(),
    }
    this.store.put('projects', item)
    this.store.audit('project.saved', { id: item.id })
    return item
  }

  task(value: unknown, existingId?: string) {
    const input = taskInput.parse(value)
    if (input.cron)
      nextOccurrences(input.cron, input.timezone)
    const agent = requireValue(this.store.get('agents', input.agentId), 'Agent not found')
    taskProjects(agent, input as Task, this.store.list('projects'))
    if (input.skills && policy(agent).skills !== null && input.skills.some(key => !policy(agent).skills!.includes(key)))
      throw new AppError(400, 'A task cannot use skills outside its agent’s access.')
    if (existingId)
      requireValue(this.store.get('tasks', existingId))
    const item: Task = {
      ...input,
      id: existingId ?? randomUUID(),
      createdAt:
        this.store.get('tasks', existingId ?? '')?.createdAt ?? Date.now(),
      nextRun:
        input.cron && input.enabled && !input.archived
          ? nextOccurrences(input.cron, input.timezone)[0]
          : null,
    }
    this.store.put('tasks', item)
    this.store.audit('task.saved', { id: item.id })
    return item
  }

  remove(kind: 'agents' | 'projects' | 'tasks', id: string) {
    requireValue(this.store.get(kind, id))
    if (kind === 'agents' && id === MAIN_AGENT_ID)
      throw new AppError(409, 'The main agent cannot be removed.')
    if (kind === 'projects' && this.store.list('agents').some(agent => policy(agent).projects?.includes(id)))
      throw new AppError(409, 'This project is assigned to an agent. Update the agent first.')
    if (
      kind !== 'tasks'
      && this.store
        .list('tasks')
        .some(t =>
          kind === 'agents' ? t.agentId === id : t.projectId === id,
        )
    ) {
      throw new AppError(
        409,
        'This item is used by a task. Update or remove that task first.',
      )
    }
    if (
      this.store
        .active()
        .some(r =>
          kind === 'tasks'
            ? r.taskId === id
            : kind === 'projects'
              ? runProjects(r).some(project => project.id === id)
              : r.snapshot.agent.id === id,
        )
    ) {
      throw new AppError(
        409,
        'This item has active work. Cancel or wait for the run first.',
      )
    }
    if (kind === 'agents')
      this.store.delete(`agent-github:${id}`)
    this.store.remove(kind, id)
    this.store.audit(`${kind}.deleted`, { id })
  }

  async agentSkills(agent: Agent) {
    const access = policy(agent)
    const available = await this.skills.list()
    for (const project of allowedProjects(agent, this.store.list('projects')))
      available.push(...await this.skills.list(project.id, project.path))
    return available.filter(skill => access.skills === null || access.skills.includes(`${skill.scope}/${skill.name}`))
  }

  async enqueue(
    taskId: string,
    trigger = 'manual',
    dedupe: string | null = null,
  ): Promise<Run> {
    const task = requireValue(
      this.store.get('tasks', taskId),
      'Task not found',
    )
    const run = await this.snapshotRun(task, trigger)
    try {
      this.store.addRun(run, dedupe)
    }
    catch (e) {
      if ((e as Error).message.includes('UNIQUE constraint')) {
        throw new AppError(
          409,
          'This task already has active work or this occurrence was already queued.',
        )
      }
      throw e
    }
    this.store.event(run.id, 'status', 'Queued')
    this.store.audit('run.queued', { id: run.id, taskId, trigger })
    return run
  }

  async snapshotRun(task: Task, trigger: string): Promise<Run> {
    if (task.archived)
      throw new AppError(409, 'Restore this archived task before running it.')
    const agent = requireValue(this.store.get('agents', task.agentId))
    const projects = taskProjects(agent, task, this.store.list('projects'))
    const project = projects.length === 1 ? projects[0] : null
    const access = policy(agent)
    const available = (await this.agentSkills(agent)).filter(skill => skill.scope === 'global' || projects.some(project => project.id === skill.scope))
    const keys = task.skills ?? available.filter(skill => skill.valid).map(skill => `${skill.scope}/${skill.name}`)
    const skills = keys.map((key) => {
      if (access.skills !== null && !access.skills.includes(key))
        throw new AppError(400, `Skill outside agent access: ${key}`)
      const selected = available.find(s => `${s.scope}/${s.name}` === key)
      if (!selected?.valid)
        throw new AppError(400, `Skill unavailable or invalid: ${key}`)
      return { name: selected.name, path: selected.path, content: selected.content }
    })
    const run: Run = {
      id: randomUUID(),
      taskId: task.id,
      projectId: project?.id ?? null,
      status: 'queued',
      trigger,
      createdAt: Date.now(),
      startedAt: null,
      finishedAt: null,
      summary: '',
      sessionId: null,
      workspace: null,
      usage: null,
      snapshot: { task, agent, project, projects, skills },
    }
    return run
  }

  async schedule(now = Date.now()) {
    for (const task of this.store.list('tasks')) {
      if (
        !task.enabled
        || task.archived
        || !task.cron
        || task.nextRun === null
        || task.nextRun > now
      ) {
        continue
      }
      try {
        await this.enqueue(task.id, 'schedule', `${task.id}:${task.nextRun}`)
      }
      catch (e) {
        if (!(e instanceof AppError && e.statusCode === 409)) {
          this.store.audit('schedule.failed', {
            taskId: task.id,
            error: (e as Error).message,
          })
        }
      }
      const current = this.store.get('tasks', task.id)
      if (current?.nextRun === task.nextRun) {
        this.store.put('tasks', {
          ...current,
          nextRun: nextOccurrences(task.cron, task.timezone, now)[0],
        })
      }
    }
  }
}

File: backend/src/service.rs
use crate::{
    auth::Auth,
    config::{Config, MAIN_AGENT_ID, id, now},
    error::{Error, Result, required},
    skills::Skills,
    store::{Store, merge},
    validation::{parse, text},
    vault::Vault,
};
use serde_json::{Value, json};
use std::{path::Path, sync::Arc};
use tokio_util::sync::CancellationToken;
#[derive(Clone)]
pub struct Service {
    pub worker: Arc<crate::worker::Worker>,
    pub mcps: Arc<crate::mcps::Mcps>,
    pub legacy_codex_login: Arc<std::sync::atomic::AtomicBool>,
    pub accounts: Arc<crate::accounts::Accounts>,
    pub connections: Arc<crate::connections::Connections>,
    pub account_login: Arc<tokio::sync::Mutex<Option<crate::connections::AccountLogin>>>,
    pub notifications: crate::notifications::Notifications,
    pub config: Config,
    pub store: Store,
    pub auth: Auth,
    pub vault: Vault,
    pub skills: Skills,
    pub http: reqwest::Client,
    pub shutdown: CancellationToken,
}
impl Service {
    pub async fn new(config: Config) -> Result<Arc<Self>> {
        let store = Store::open(&config.data_dir)?;
        let vault = Vault::new(store.clone(), &config.data_dir)?;
        let service = Arc::new(Self {
            worker: Default::default(),
            mcps: Default::default(),
            legacy_codex_login: Default::default(),
            accounts: Default::default(),
            connections: Default::default(),
            account_login: Default::default(),
            notifications: Default::default(),
            auth: Auth::new(store.clone(), config.public_url.clone()),
            skills: Skills {
                config: config.clone(),
            },
            http: reqwest::Client::builder()
                .redirect(reqwest::redirect::Policy::none())
                .timeout(std::time::Duration::from_secs(20))
                .build()
                .map_err(Error::internal)?,
            config,
            store,
            vault,
            shutdown: CancellationToken::new(),
        });
        for mut agent in service.store.list("agents").await? {
            if agent["access"].get("mcps").is_none() {
                agent["access"] = policy(&agent);
                service.store.put("agents", agent).await?;
            }
        }
        if service.store.get("agents", MAIN_AGENT_ID).await?.is_none() {
            let mut agent = parse(
                "agent",
                json!({
                "name":"Main agent","description":"Your default agent, with access to every registered project, skill, and shared connection."}
                ),
            )?;
            agent["id"] = MAIN_AGENT_ID.into();
            agent["createdAt"] = now().into();
            service.store.put("agents", agent).await?;
        }
        Ok(service)
    }
    pub async fn get(&self, kind: &str, id: &str) -> Result<Value> {
        required(self.store.get(kind, id).await?, "Record not found")
    }
    pub async fn agent(&self, mut input: Value, existing_id: Option<&str>) -> Result<Value> {
        let existing = if let Some(id) = existing_id {
            Some(self.get("agents", id).await?)
        } else {
            None
        };
        if input.get("access").is_none()
            && let Some(existing) = &existing
        {
            input["access"] = policy(existing);
        }
        let mut agent = parse("agent", input)?;
        agent["id"] = existing_id.map(str::to_owned).unwrap_or_else(id).into();
        agent["createdAt"] = existing
            .as_ref()
            .map(|v| v["createdAt"].clone())
            .unwrap_or_else(|| now().into());
        let access = policy(&agent);
        if !access["projects"].is_null() && access["github"] == true {
            return Err(Error::bad(
                "Shared GitHub credentials require access to all projects. Disable the GitHub connection for an agent with selected projects.",
            ));
        }
        if agent["id"] == MAIN_AGENT_ID
            && (!access["projects"].is_null()
                || !access["skills"].is_null()
                || access["github"] != true
                || !access["mcps"].is_null()
                || access["mcpTools"]
                    .as_object()
                    .is_some_and(|v| !v.is_empty()))
        {
            return Err(Error::bad(
                "The main agent always has access to all resources. Create another agent for restricted access.",
            ));
        }
        for kind in ["projects", "mcps"] {
            for value in access[kind].as_array().into_iter().flatten() {
                self.get(kind, value.as_str().unwrap_or("")).await?;
            }
        }
        if let Some(tools) = access["mcpTools"].as_object() {
            for key in tools.keys() {
                self.get("mcps", key).await?;
                if !allowed(&access["mcps"], key) {
                    return Err(Error::bad(
                        "Tool permissions require access to the MCP connection.",
                    ));
                }
            }
        }
        self.store.save("agents", agent, "agent.saved").await
    }
    pub async fn project(&self, input: Value, existing: Option<&str>) -> Result<Value> {
        let mut project = parse("project", input)?;
        let actual = crate::skills::workspace(
            Path::new(text(&project, "path")),
            &self.config.workspace_roots,
        )
        .await?;
        let previous = if let Some(id) = existing {
            Some(self.get("projects", id).await?)
        } else {
            None
        };
        let args = [
            "-C",
            actual.to_str().unwrap_or(""),
            "config",
            "--get",
            "remote.origin.url",
        ]
        .map(str::to_owned);
        let origin = crate::process::bounded_output(
            crate::process::command("git", &args, &std::env::vars().collect(), None),
            std::time::Duration::from_secs(3),
            8000,
        )
        .await
        .ok()
        .filter(|output| output.success)
        .map(|output| output.stdout.trim().to_owned())
        .unwrap_or_default();
        project["origin"] = normalize_origin(&origin).into();
        project["path"] = actual.to_string_lossy().into_owned().into();
        project["id"] = existing.map(str::to_owned).unwrap_or_else(id).into();
        project["createdAt"] = previous
            .map(|v| v["createdAt"].clone())
            .unwrap_or_else(|| now().into());
        self.store.save("projects", project, "project.saved").await
    }
    pub async fn task(&self, input: Value, existing: Option<&str>) -> Result<Value> {
        let mut task = parse("task", input)?;
        let agent = self.get("agents", text(&task, "agentId")).await?;
        task_projects(&agent, &task, &self.store.list("projects").await?)?;
        if let Some(cron) = task["cron"].as_str() {
            next_occurrences(cron, text(&task, "timezone"), now(), 1)?;
        }
        task["id"] = existing.map(str::to_owned).unwrap_or_else(id).into();
        task["createdAt"] = if let Some(id) = existing {
            self.get("tasks", id).await?["createdAt"].clone()
        } else {
            now().into()
        };
        task["nextRun"] =
            if task["enabled"] == true && task["archived"] != true && task["cron"].is_string() {
                next_occurrences(text(&task, "cron"), text(&task, "timezone"), now(), 1)?[0].into()
            } else {
                Value::Null
            };
        self.store.save("tasks", task, "task.saved").await
    }
    pub async fn remove(&self, kind: &str, id: &str) -> Result<()> {
        let (kind, id) = (kind.to_owned(), id.to_owned());
        self.store
            .transaction(move |db| {
                required(db.get(&kind, &id)?, "Record not found")?;
                if kind == "agents" && id == MAIN_AGENT_ID {
                    return Err(Error::new(409, "The main agent cannot be removed."));
                }
                if kind == "projects"
                    && db.list("agents")?.iter().any(|a| {
                        !policy(a)["projects"].is_null() && allowed(&policy(a)["projects"], &id)
                    })
                {
                    return Err(Error::new(
                        409,
                        "This project is assigned to an agent. Update the agent first.",
                    ));
                }
                if kind != "tasks"
                    && db.list("tasks")?.iter().any(|t| {
                        t[if kind == "agents" {
                            "agentId"
                        } else {
                            "projectId"
                        }] == id
                    })
                {
                    return Err(Error::new(
                        409,
                        "This item is used by a task. Update or remove that task first.",
                    ));
                }
                if db.active()?.iter().any(|r| match kind.as_str() {
                    "tasks" => r["taskId"] == id,
                    "projects" => run_projects(r).iter().any(|p| p["id"] == id),
                    _ => r["snapshot"]["agent"]["id"] == id,
                }) {
                    return Err(Error::new(
                        409,
                        "This item has active work. Cancel or wait for the run first.",
                    ));
                }
                if kind == "agents" {
                    db.delete(&format!("agent-github:{id}"))?;
                }
                db.remove(&kind, &id)?;
                db.audit(
                    &format!("{kind}.deleted"),
                    &json!({
                    "id":id}
                    ),
                )
            })
            .await
    }
    pub async fn agent_skills(&self, agent: &Value) -> Result<Vec<Value>> {
        let access = policy(agent);
        let mut available = self.skills.list("global", None).await?;
        for project in self.store.list("projects").await? {
            if allowed(&access["projects"], text(&project, "id")) {
                available.extend(
                    self.skills
                        .list(
                            text(&project, "id"),
                            Some(Path::new(text(&project, "path"))),
                        )
                        .await?,
                );
            }
        }
        available.retain(|s| {
            allowed(
                &access["skills"],
                &format!("{}/{}", text(s, "scope"), text(s, "name")),
            )
        });
        Ok(available)
    }
    pub async fn snapshot(&self, task: Value, trigger: &str) -> Result<Value> {
        if task["archived"] == true {
            return Err(Error::new(
                409,
                "Restore this archived task before running it.",
            ));
        }
        let agent = self.get("agents", text(&task, "agentId")).await?;
        let projects = task_projects(&agent, &task, &self.store.list("projects").await?)?;
        let project = if projects.len() == 1 {
            projects[0].clone()
        } else {
            Value::Null
        };
        let available = self
            .agent_skills(&agent)
            .await?
            .into_iter()
            .filter(|s| s["scope"] == "global" || projects.iter().any(|p| p["id"] == s["scope"]))
            .collect::<Vec<_>>();
        let keys = task["skills"].as_array().cloned().unwrap_or_else(|| {
            available
                .iter()
                .filter(|s| s["valid"] == true)
                .map(|s| format!("{}/{}", text(s, "scope"), text(s, "name")).into())
                .collect()
        });
        let mut skills = Vec::new();
        for key in keys {
            let key = key.as_str().unwrap_or("");
            if !allowed(&policy(&agent)["skills"], key) {
                return Err(Error::bad(format!("Skill outside agent access: {key}")));
            }
            let s = available
                .iter()
                .find(|s| {
                    format!("{}/{}", text(s, "scope"), text(s, "name")) == key && s["valid"] == true
                })
                .ok_or_else(|| Error::bad(format!("Skill unavailable or invalid: {key}")))?;
            skills.push(json!({
            "name":s["name"],"path":s["path"],"content":s["content"]}
            ));
        }
        Ok(json!({
        "id":id(),"taskId":task["id"],"projectId":project["id"],"status":"queued","trigger":trigger,"createdAt":now(),"startedAt":null,"finishedAt":null,"summary":"","sessionId":null,"workspace":null,"usage":null,"snapshot":{
        "task":task,"agent":agent,"project":project,"projects":projects,"skills":skills}
        }
        ))
    }
    pub async fn enqueue(
        &self,
        task_id: &str,
        trigger: &str,
        dedupe: Option<String>,
    ) -> Result<Value> {
        let run = self
            .snapshot(self.get("tasks", task_id).await?, trigger)
            .await?;
        self.store
            .transaction(move |db| {
                db.add_run(&run, dedupe.as_deref())?;
                db.event(text(&run, "id"), "status", "Queued", None)?;
                db.audit(
                    "run.queued",
                    &json!({
                    "id":run["id"],"taskId":run["taskId"],"trigger":run["trigger"]}
                    ),
                )?;
                Ok(run)
            })
            .await
    }
    pub async fn schedule(&self) -> Result<()> {
        for task in self.store.list("tasks").await? {
            if task["enabled"] != true
                || task["archived"] == true
                || task["cron"].is_null()
                || task["nextRun"].as_i64().is_none_or(|time| time > now())
            {
                continue;
            }
            if let Err(error) = self
                .enqueue(
                    text(&task, "id"),
                    "schedule",
                    Some(format!("{}:{}", text(&task, "id"), task["nextRun"])),
                )
                .await
                && error.status != 409
            {
                self.store
                    .audit(
                        "schedule.failed",
                        json!({
                        "taskId":task["id"],"error":error.message}
                        ),
                    )
                    .await?;
            }
            let next = next_occurrences(text(&task, "cron"), text(&task, "timezone"), now(), 1)?[0];
            self.store
                .write(move |db| {
                    if let Some(mut current) = db.get("tasks", text(&task, "id"))?
                        && current["nextRun"] == task["nextRun"]
                    {
                        current["nextRun"] = next.into();
                        db.put("tasks", &current)?;
                    }
                    Ok(())
                })
                .await?;
        }
        Ok(())
    }
}
pub fn policy(agent: &Value) -> Value {
    let mut value = json!({
    "projects":null,"skills":null,"mcps":null,"mcpTools":{
    }
    ,"github":true,"sandbox":"yolo"}
    );
    merge(&mut value, &agent["access"]);
    if agent["id"] != MAIN_AGENT_ID
        && agent["access"].is_object()
        && agent["access"].get("mcps").is_none()
        && (!value["projects"].is_null()
            || !value["skills"].is_null()
            || value["github"] != true
            || value["sandbox"] != "yolo")
    {
        value["mcps"] = json!([]);
    }
    value
}
pub fn allowed(scope: &Value, id: &str) -> bool {
    scope.is_null() || scope.as_array().is_some_and(|a| a.iter().any(|v| v == id))
}
pub fn isolated(agent: &Value) -> bool {
    let a = policy(agent);
    !a["projects"].is_null()
        || !a["skills"].is_null()
        || !a["mcps"].is_null()
        || a["github"] != true
        || a["sandbox"] != "yolo"
        || a["mcpTools"].as_object().is_some_and(|a| !a.is_empty())
}
pub fn task_projects(agent: &Value, task: &Value, projects: &[Value]) -> Result<Vec<Value>> {
    let a = policy(agent);
    let available = projects
        .iter()
        .filter(|p| allowed(&a["projects"], text(p, "id")))
        .cloned()
        .collect::<Vec<_>>();
    if !task["projectId"].is_null() {
        return available
            .into_iter()
            .find(|p| p["id"] == task["projectId"])
            .map(|p| vec![p])
            .ok_or_else(|| Error::bad("This project is unavailable to the selected agent."));
    }
    Ok(available)
}
pub fn run_projects(run: &Value) -> Vec<Value> {
    run["snapshot"]["projects"]
        .as_array()
        .cloned()
        .unwrap_or_else(|| {
            if run["snapshot"]["project"].is_object() {
                vec![run["snapshot"]["project"].clone()]
            } else {
                vec![]
            }
        })
}
pub fn next_occurrences(pattern: &str, zone: &str, time: i64, count: usize) -> Result<Vec<i64>> {
    use chrono::{Offset, TimeZone};
    use std::str::FromStr;
    if pattern.split_whitespace().count() != 5 {
        return Err(Error::bad(
            "Enter a valid five-field cron expression and IANA timezone.",
        ));
    }
    let zone = chrono_tz::Tz::from_str(zone)
        .map_err(|_| Error::bad("Enter a valid five-field cron expression and IANA timezone."))?;
    let cron = croner::Cron::from_str(pattern)
        .map_err(|_| Error::bad("Enter a valid five-field cron expression and IANA timezone."))?;
    let mut current = zone
        .timestamp_millis_opt(time)
        .single()
        .ok_or_else(|| Error::bad("Invalid schedule time"))?;
    let mut result = Vec::new();
    for _ in 0..count {
        let mut next = cron
            .find_next_occurrence(&current, false)
            .map_err(|_| Error::bad("No future schedule occurrence."))?;
        // cron-parser preserves the scheduled minute when a fixed wall-clock
        // time falls in the spring gap. Croner clamps it to the gap's end.
        // Resolve that missing time using the offset immediately before the gap.
        if !cron.is_time_matching(&next).map_err(Error::internal)? {
            let before = next - chrono::Duration::hours(3);
            let offset = before.offset().fix();
            let shifted = cron
                .find_next_occurrence(&current.with_timezone(&offset), false)
                .map_err(Error::internal)?
                .with_timezone(&zone);
            if shifted >= next && shifted - next < chrono::Duration::hours(3) {
                next = shifted;
            }
        }
        current = next;
        result.push(current.timestamp_millis());
    }
    Ok(result)
}
fn normalize_origin(remote: &str) -> String {
    if let Ok(mut url) = url::Url::parse(remote)
        && ["http", "https", "ssh", "git"].contains(&url.scheme())
    {
        let _ = url.set_username("");
        let _ = url.set_password(None);
        url.set_query(None);
        url.set_fragment(None);
        return url.to_string();
    }
    if let Some((host, path)) = remote.split_once(':') {
        let host = host.rsplit('@').next().unwrap_or(host);
        if !host.is_empty() && !path.is_empty() && !host.contains('/') {
            return format!("{host}:{path}");
        }
    }
    String::new()
}
